### PR TITLE
[RDY] *: Partial string handling refactoring

### PR DESCRIPTION
**Pull request by ZyX-I**
Main points:

- Replace `char_u` with `char` in some cases.
- Remove `str[len] = NUL` hack in some cases when `str` may be considered
  `const`.

Part of #5119

### CODE REVIEW

**Comment by ZyX-I** [comment]
QB timed out on “test/functional/clipboard/clipboard_provider_spec.lua @ 287: clipboard usage with clipboard=unnamed is updated on global changes”. AFAIR it is not the first time I see timeout somewhere there. Seems unrelated.

PR should be ready now as other builds passed.

File: src/nvim/hashtab.c [hunk 1]
```diff
@@ -375,14 +401,43 @@ hash_T hash_hash(char_u *key)
...
+{
+  if (len == 0) {
+    return 0;
+  }
+
+  hash_T hash = *(uint8_t *)key;
+  const uint8_t *end = (uint8_t *)key + len;
+
+  const uint8_t *p = (const uint8_t *)key + 1;
+  while (p < end) {
```
Comment (justinmk) [inline src/nvim/hashtab.c]: why not do a sanity check like `assert(*p !- NUL)` here?

File: src/nvim/message.c [hunk 2]
```diff
@@ -382,39 +383,40 @@ static int other_sourcing_name(void)
   return FALSE;
 }
 
-/// Get the message about the source, as used for an error message.
-/// Returns an allocated string with room for one more character.
-/// Returns NULL when no message is to be given.
-static char_u *get_emsg_source(void)
+/// Get the message about the source, as used for an error message
+///
+/// @return [allocated] String with room for one more characters. NULL when no
```
Comment (justinmk) [inline src/nvim/message.c]: typo: characters should be singular

File: src/nvim/hashtab.c [hunk 3]
```diff
@@ -82,22 +82,42 @@ void hash_clear_all(hashtab_T *ht, unsigned int off)
...
+/// @param[in]  key  Key of the looked-for item. Must not be NULL.
+/// @param[in]  len  Key length.
+///
+/// @return Pointer to the hash item corresponding to the given key.
+///         If not found, then return pointer to the empty item that would be
+///         used for that key.
+///
+///         @warning Returned pointer becomes invalid as soon as the hash table
+///                  is changed in any way.
+hashitem_T *hash_find_len(hashtab_T *ht, const char *key, const size_t len)
```
Comment (jamessan) [inline src/nvim/hashtab.c]: Was this just added as an optimization when `len` is pre-computed, or is there an expectation of having keys that contain `NUL`?

Comment (ZyX-I) [inline src/nvim/hashtab.c]: It is not an optimization: it is needed to be able to use strings like `bar)` (part of `string(foo.bar)`) without either copying string somewhere or modifying input string. Though I do not remember where exactly this is used in this PR: `dict_find` (which is the function to receive `bar)` in the example) is still copying string over (but `tv_dict_find` from #5119  no longer does this).

Comment (ZyX-I) [inline src/nvim/hashtab.c]: Though with my description this may be seen as an optimization. But AFAIR I was actually avoiding `*p = NUL` hack somewhere.

**Pull request merged**